src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/SolutionService.cs

```diff
@@ -0,0 +1,102 @@
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    /// <inheritdoc cref="ISolutionService"/>
+    [Export(typeof(ISolutionService))]
+    internal sealed class SolutionService : ISolutionService, IVsSolutionEvents, IVsPrioritizedSolutionEvents, IDisposable
+    {
+        private const int StateUninitialized = 0;
+        private const int StateListening = 1;
+        private const int StateDisposed = 2;
+
+        private readonly IVsUIService<IVsSolution> _solution;
+        private readonly JoinableTaskContext _joinableTaskContext;
+
+        private int _state = StateUninitialized;
+        private uint _cookie = VSConstants.VSCOOKIE_NIL;
+        
+        /// <inheritdoc />
+        public bool IsSolutionClosing { get; private set; }
+
+        [ImportingConstructor]
+        public SolutionService(IVsUIService<SVsSolution, IVsSolution> solution, JoinableTaskContext joinableTaskContext)
+        {
+            _solution = solution;
+            _joinableTaskContext = joinableTaskContext;
+        }
+
+        public void StartListening()
+        {
+            Assumes.True(_joinableTaskContext.IsOnMainThread, "Must be called on the UI thread.");
+
+            if (Interlocked.CompareExchange(ref _state, StateListening, StateUninitialized) != StateUninitialized)
+            {
+                return;
+            }
+
+            IVsSolution? solution = _solution.Value;
+            Assumes.Present(solution);
+
+            Verify.HResult(solution.AdviseSolutionEvents(this, out _cookie));
+        }
+
+        public int OnAfterOpenSolution(object pUnkReserved, int fNewSolution)                                 => UpdateClosing(false);
+        public int OnBeforeCloseSolution(object pUnkReserved)                                                 => UpdateClosing(true);
+        public int OnAfterOpenProject(IVsHierarchy pHierarchy, int fAdded)                                    => HResult.OK;
+        public int OnQueryCloseProject(IVsHierarchy pHierarchy, int fRemoving, ref int pfCancel)              => HResult.OK;
+        public int OnBeforeCloseProject(IVsHierarchy pHierarchy, int fRemoved)                                => HResult.OK;
+        public int OnAfterLoadProject(IVsHierarchy pStubHierarchy, IVsHierarchy pRealHierarchy)               => HResult.OK;
+        public int OnQueryUnloadProject(IVsHierarchy pRealHierarchy, ref int pfCancel)                        => HResult.OK;
+        public int OnBeforeUnloadProject(IVsHierarchy pRealHierarchy, IVsHierarchy pStubHierarchy)            => HResult.OK;
+        public int OnQueryCloseSolution(object pUnkReserved, ref int pfCancel)                                => HResult.OK;
+        public int OnAfterCloseSolution(object pUnkReserved)                                                  => HResult.OK;
+
+        public int PrioritizedOnAfterOpenSolution(object pUnkReserved, int fNewSolution)                      => UpdateClosing(false);
```

Comment:
I still think this should pick one Before and one After. Otherwise, someone will debug this and be confused why we're setting a property twice with the same value. I see no value in this other than debugging confusion.

Reply:
My goal was to set the flag as early as possible and increase confidence in correctness in cases I might not have tested for. I will add a comment to address any potential confusion.